features/device_key/source/DeviceKey.h

```diff
@@ -0,0 +1,142 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_DEVICEKEY_H
+#define MBED_DEVICEKEY_H
+
+#include "stddef.h"
+#include "stdint.h"
+#include "platform/NonCopyable.h"
+
+#if (NVSTORE_ENABLED)
+
+namespace mbed {
+/** \addtogroup drivers */
+
+#define DEVICE_KEY_16BYTE 16
+#define DEVICE_KEY_32BYTE 32
+
+enum DeviceKeyStatus {
+    DEVICEKEY_SUCCESS                     =  0,
+    DEVICEKEY_INVALID_KEY_SIZE            = -1,
+    DEVICEKEY_INVALID_KEY_TYPE            = -2,
+    DEVICEKEY_SAVE_FAILED                 = -3,
+    DEVICEKEY_ALREADY_EXIST               = -4,
+    DEVICEKEY_NOT_FOUND                   = -5,
+    DEVICEKEY_READ_FAILED                 = -6,
+    DEVICEKEY_NVSTORE_UNPREDICTABLE_ERROR = -7,
+    DEVICEKEY_ERR_CMAC_GENERIC_FAILURE    = -8,
+    DEVICEKEY_BUFFER_TO_SMALL             = -9,
+    DEVICEKEY_NO_KEY_INJECTED             = -10,
+    DEVICEKEY_INVALID_PARAM               = -11,
+    DEVICEKEY_TRNG_ERROR                  = -12,
+};
+
+/** Use this singleton if you need to derive a new key from the device root of trust.
+ *
+ * @note Synchronization level: Thread safe
+ * @ingroup drivers
+ */
+
+class DeviceKey : private mbed::NonCopyable<DeviceKey> {
+public:
+
+    /**
+     * @brief As a singleton, return the single instance of the class.
+     *        Reason for this class being a singleton is the following:
+     *        - Ease the use for users of this class not having to coordinate instantiations.
+     *        - Lazy instantiation of internal data (which we can't achieve with simple static classes).
+     *
+     * @returns Singleton instance reference.
+     */
+    static DeviceKey& get_instance()
+    {
+        // Use this implementation of singleton (Meyer's) rather than the one that allocates
+        // the instance on the heap, as it ensures destruction at program end (preventing warnings
+        // from memory checking tools such as valgrind).
+        static DeviceKey instance;
+        return instance;
+    }
+
+    ~DeviceKey();
+
+    /** Derive a new key based on the salt string. key type can be with values 16 bytes and 32 bytes
+     * @param isalt input buffer used to create the new key. Same input will generate always the same key
+     * @param isalt_size size of the data in salt buffer
+     * @param output buffer to receive the derived key. Size must be 16 bytes or 32 bytes
+     *               according to the ikey_type parameter
+     * @param ikey_type type of the required key. Type must be 16 bytes or 32 bytes.
+     * @return 0 on success, negative error code on failure
+     */
+    int generate_derived_key(const unsigned char *isalt, size_t isalt_size, unsigned char *output, uint16_t ikey_type);
```

Comment:
@mazimkhan 
This was discussed earlier in this review and decided to be changed when the MbedErrorStatus will be introduce because then all features will use only one type.